src/main/resources/avro/genotypephenotypemethods.avdl

```diff
@@ -0,0 +1,159 @@
+Disease Ontology (DO): http://purl.obolibrary.org/obo/DOID_4
+
+Feature:
+Sequence Ontology (SO): http://www.sequenceontology.org/browser/
+
+Evidence:
+Evidence Code Ontology (ECO): http://www.ontobee.org/browser/index.php?o=ECO
+Ontology for Biomedical Investigations (OBI): http://purl.obofoundry.org/obo/obi/browse
+*/
+protocol GenotypePhenotypeMethods {
+
+
+import idl "genotypephenotype.avdl";
+import idl "methods.avdl";
+import idl "common.avdl";
+
+/******************  /genotypephenotype/search  *********************/
+
+
+
+/**
+Evidence for the phenotype association.
+*/
+record EvidenceQuery {
+  /** ECO or OBI is recommended */
+  array<org.ga4gh.models.OntologyTerm> evidenceType;
+}
+
+
+/**
+The feature collection to search for.  One or more features (variants,
+genes, etc) can be specified.  The idea here is that if a
+query is for a gene, then any alleles to that gene for which
+there are annotation records would be returned.
+*/
+record GenomicFeatureQuery {
+  array<org.ga4gh.models.Feature> features;
+}
+
+
+/**
+One or more phenotypes can be queried together.
+*/
+record PhenotypeQuery {
+  array<org.ga4gh.models.PhenotypeInstance> phenotypes;
+}
+
+/**
+One or more ontology terms can be queried together.
+*/
+record OntologyTermQuery {
+  array<org.ga4gh.models.OntologyTerm> terms;
+}
+
+/**
+One or more ids can be queried together.  Generally used for instances
+of a particular class of object (e.g. a specific gene or SNP).
+*/
+record ExternalIdentifierQuery {
+  array<org.ga4gh.models.ExternalIdentifier> ids;
+}
+
+/**
+This request maps to the body of `POST /genotypephenotype/search` as JSON.
+
+The goal here is to allow users to query using one or more of
+Genotype, Phenotype, Environment, and Evidence.
+
+A query using one of the above items is to mean, by default,
+that the remainder of the query is as a "wildcard", such
+that all matches to just that query term would come back.
+Combinations of the above are to act like AND rather than OR.
+
+The "genotype" part of the query methods can be one or more
+genomic features.  Associations can be made at many
+levels of granularity (from whole genotypes down to individual
+SNVs), but users may use these methods with partial or
+inexact information.  Therefore, the query methods must be
+able to support query of some or all of the associated features.
+Furthermore, use of the relationships between genomic features
+means that when querying for a gene, any variants to that
+gene are also returned.  For example, a query with
+BRCA2 would mean that in addition to any direct associations
+to the BRCA2, all associations to sequence variants of BRCA2 would also
+be returned.  Similarly, queries with OntologyTerms should perform
+the subclass closure.
+
+Each query can be made against a string, an array of external
+identifers (such as for gene or SNP ids), ontology term ids, or
+full feature/phenotype/evidence objects.
+*/
+record SearchGenotypePhenotypeRequest {
```

Comment:
as usual, @reece says it far better than I can.

The subject of the queries is not defined.  What does string
match in one of these queries?  What happens when the objects
evolve?

The lack of precision in the APIs is a concern.  GA4GH took on a
very big challenge that requires rigorous definition: defining
an interoperable API.  It's much easier for Facebook to define
an API, because their is one implementation.

Reece Hart <notifications@github.com> writes:
> In src/main/resources/avro/genotypephenotypemethods.avdl:
> 
> > +SNVs), but users may use these methods with partial or
> > +inexact information.  Therefore, the query methods must be
> > +able to support query of some or all of the associated features.
> > +Furthermore, use of the relationships between genomic features
> > +means that when querying for a gene, any variants to that
> > +gene are also returned.  For example, a query with
> > +BRCA2 would mean that in addition to any direct associations
> > +to the BRCA2, all associations to sequence variants of BRCA2 would also
> > +be returned.  Similarly, queries with OntologyTerms should perform
> > +the subclass closure.
> > +
> > +Each query can be made against a string, an array of external
> > +identifers (such as for gene or SNP ids), ontology term ids, or
> > +full feature/phenotype/evidence objects.
> > +*/
> > +record SearchGenotypePhenotypeRequest {
> 
> If I understand this correctly, I think we should be concerned about clashing
> of unscoped identifiers. For example, I read this as supporting something like
> { 'phenotype': 'FH' }, in which I think it's unclear whether that's FH the gene
> (via "ExternalIdentifierQuery") or Familial Hypercholesterolemia (via
> "PhenotypeQuery"). Is that (or something like it) a valid concern here?
> 
> [Larger conversation: We (GA4GH) are using unions as a way to avoid important
> conversations regarding precision of meaning. This is directly analogous to
> strong v. weak typing tradeoffs in programming languages. Anyone else have
> similar concerns?]
> 
> —
> Reply to this email directly or view it on GitHub.*
>